### ISSUE TITLE
Review maximum cardinality search algorithm in `complete_to_chordal_graph`

### ISSUE THREAD

**Issue by richardrl** [networkx/networkx#6873]
```
(Pdb) G.add_nodes_from([1, 2, 3])
(Pdb) G.add_edges_from([(1,2), (1,3)])
(Pdb) G
<networkx.classes.graph.Graph object at 0x7fecb091e610>
(Pdb) nx.complete_to_chordal_graph(G)
(<networkx.classes.graph.Graph object at 0x7fecb092fb50>, {1: 0, 2: 0, 3: 0})

```
Shouldn't a PEO here be 2, 1, 3?

**Comment by dschult** [comment]
The doc_string says this returns an elimination ordering -- not a perfect elimination ordering.
From the code it is clear that if the graph is chordal, this function will just return a dict keyed by node to the value 0 for all nodes. So, that is not related to perfect elimination ordering at all.  Here is [a comment about the return value `alpha`](https://github.com/networkx/networkx/pull/3353#discussion_r284670279) and what it might be used for. 

I find the "elimination ordering" in this function suspicious. And the handling of chordal graphs in this function does not produce a perfect elimination ordering.  Perhaps you can figure out how the algorithm is supposed to produce a perfect elimination ordering -- or whether that is even the intent.
 
The algorithm and a description of this elimination ordering is described in the papers:
The original algorithm
"[Maximum cardinality search for computing minimal triangulations of graphs](http://www.ii.uib.no/~pinar/MCSM-r.pdf)"
and a later article comparing it to another algorithm
"[Lex M versus MCS-M](https://www.sciencedirect.com/science/article/pii/S0012365X05006084)"

I'd love some feedback/direction on this topic and function as I don't know the chordal_graph literature well at all.

**Comment by nihalgeorge01** [comment]
I just read the code. There is a preliminary check if the graph is chordal in complete_to_chordal_graph(), and if so it returns immediately. [source](https://github.com/networkx/networkx/tree/main/networkx/algorithms/chordal.py)
```
H = G.copy()
alpha = {node: 0 for node in H}
if nx.is_chordal(H):
    return H, alpha
```
But before the check, alpha is set to all zeros. If this check is commented out and the function is run to completion, we do get a PEO above as {1:3, 2:2, 3:1} which is vertices 3, 2, 1. The edge set remains the same (which is what we want since G is chordal). So it seems like running the MCS-M code for an already chordal graph will keep the edges as-is, but populates the alpha correctly to give a PEO. 

This paper [Algorithmic Aspects of Vertex Elimination on Directed Graphs, page 5 lower](http://i.stanford.edu/pub/cstr/reports/cs/tr/75/531/CS-TR-75-531.pdf) says that a minimal EO in any perfect elimination graph will also be a PEO (PE graph <-> a graph having a PEO <-> a chordal graph). The first MCS-M paper you linked shows that MCS-M computes a minimal EO. So MCS-M should compute a PEO if given a chordal graph. In non-chordal cases, it computes a minimal EO as advertised.

The tests in networkx/algorithms/tests/test_chordal.py still work after this modification, if we omit the explicit check for alpha equaling all zeros for an already chordal graph. [tests source](https://github.com/networkx/networkx/tree/main/networkx/algorithms/tests/test_chordal.py)
```
if nx.is_chordal(G):
    assert G.number_of_edges() == H.number_of_edges()
    # assert set(a.values()) == {0} # commented out
else:
    assert len(set(a.values())) == H.number_of_nodes()
```

**Comment by dschult** [comment]
Great!  Thanks for digging through this and for finding the references, etc.

So, it seems that we have many functions that could all use this underlying code:
- make_perfect_elimination_graph  : return the graph created in current code
- minimal_elimination_order : return the order given by alpha from the current code
- perfect_elimination_order : if not is_chordal raise, else return MEO given by alpha
- is_perfect_elimination_graph : is_chordal   (no new function needed... perhaps add a sentence to docstring.

Maybe the change to `complete_to_chordal_graph` would need to be deprecated..?? The old behavior was to return alpha with all zeros. Can we just change that? It is not a useful return value. Do we need to deprecate?  

If we do, let's copy the existing code to a private function, remove the is_chordal check and make a new function with the old name and doc_string that does the check and returns alpha with zeros. Then the other functions can call the private function. I suppose we could make a new public function that returns both the graph and alpha.  What would a better name for that function be?  The current function certainly does much more than making a graph chordal. :}

Thoughts?

**Comment by rossbar** [comment]
Thanks @dschult , that helps clarify the landscape here around elimination order for chordal graphs. I'm still struggling to get the full picture (esp. the difference between a perfect elimination graph and a perfect elimination ordering), but overall the summary above seems like a step in the right direction to me!

> Maybe the change to complete_to_chordal_graph would need to be deprecated..?? The old behavior was to return alpha with all zeros. Can we just change that? It is not a useful return value. Do we need to deprecate?

Sounds more like a defect than a behavior change to me, so I'd think a deprecation is not necessary.

**Comment by dschult** [comment]
See also #7052

**Comment by akshitasure12** [comment]
Hi, I'm Akshita Sure, I'm planning to look into this issue.
Hope this issue still needs to be addressed!

**Comment by dschult** [comment]
Yes, this issue still needs to be addressed.  Thanks for looking into it!

**Comment by Luis-Varona** [comment]
If this issue still needs to be addressed, I'd love to take it. 🙂

**Comment by gulshan-123** [comment]
Hi @Luis-Varona, are you working on this issue?

**Comment by amcandio** [comment]
> Hi @Luis-Varona, are you working on this issue?

If you are interested feel free to send PR. "Worst" case we have two PRs at the same time and we work to consolidate into one (GitHub supports co-authors)

**Comment by Luis-Varona** [comment]
@gulshan-123 @amcandio I'm working on it, but feel free to start your own PR if you'd like. Might take me a few more days 🙂

**Comment by amcandio** [comment]
> @gulshan-123 @amcandio I'm working on it, but feel free to start your own PR if you'd like. Might take me a few more days 🙂

Great! If you want you can also share draft PR to get early feedback.

**Comment by Luis-Varona** [comment]
@amcandio, I just put up a draft PR at #8024; all that's left, pending approval, is to make docstrings. I also have some ideas about potentially better naming conventions for the new functions proposed by @dschult?

**Comment by Luis-Varona** [comment]
> Can we settle some terminology here? What does the term "perfect elimination graph" refer to in this module? The term "perfect elimination order(ing)" is defined by [Wikipedia's chordal graph entry](https://en.wikipedia.org/wiki/Chordal_graph), and many papers, as an order of the nodes such that a clique is formed by each node and its neighbors that are later in the ordering. A graph with a perfect elimination ordering is a chordal graph. But what is a perfect elimination graph?
> 
> It seems that "perfect elimination graph" may just be another term for "chordal graph". If so, let's make that clear. Perhaps only using `chordal_graph` in function names and adding the other name in the doc_strings consistently. If not, what is a perfect elimination graph? BTW, it looks like "decomposable graphs", "triangulated graphs", "rigid circuit graphs" are other names for "chordal graphs".

@dschult Indeed; I'll conduct some research into this topic. Will keep you updated. 🙂

**Comment by Luis-Varona** [comment]
@dschult Do you think we still need to rename the functions/clarify terminology, or do the current naming conventions in #8024 seem satisfactory to you?

**Comment by Luis-Varona** [comment]
> Can we settle some terminology here? What does the term "perfect elimination graph" refer to in this module? The term "perfect elimination order(ing)" is defined by [Wikipedia's chordal graph entry](https://en.wikipedia.org/wiki/Chordal_graph), and many papers, as an order of the nodes such that a clique is formed by each node and its neighbors that are later in the ordering. A graph with a perfect elimination ordering is a chordal graph. But what is a perfect elimination graph?
> 
> It seems that "perfect elimination graph" may just be another term for "chordal graph". If so, let's make that clear. Perhaps only using `chordal_graph` in function names and adding the other name in the doc_strings consistently. If not, what is a perfect elimination graph? BTW, it looks like "decomposable graphs", "triangulated graphs", "rigid circuit graphs" are other names for "chordal graphs".

@dschult Just following up on terminology in #8024. Would you still like to do this before merging, or are the changes @amcandio requested (which I've already made) sufficient?

**Comment by Luis-Varona** [comment]
@dschult Thanks again for the review; I think everything's ready to go in #8024 if you agree.

**Comment by rossbar** [comment]
I think there's still some work to do here to settle on terminology and behavior. I definitely agree that the interpretation of `alpha` as currently defined is hard to understand, and I don't see how the current behavior when the input graph G is chordal makes sense.

My understanding of perfect elimination ordering is that it is a sequence of nodes that, if the nodes were to be removed from `G` in that given order, the resultant graph would still be chordal at each step. Even if `alpha` is not intended to give a *perfect* elimination ordering, I don't understand how the dict is to be interpreted as an elimination ordering at all. Based on context, I suspect it's a dict keyed by node with values corresponding to the order in which the node should be removed (1-indexed). Assuming this is the correct interpretation, this already feels like a clunky way to represent this information... why not a sequence of nodes? That feels much more natural. Furthermore, the case when `G` is chordal then makes even less sense, unless values of `0` in the `alpha` dict are intended as some form of "dummy" value indicating that no removal is needed. At least, that's the only interpretation I can come up with (again, from context since this isn't documented), but ironically the only time the elimination ordering *could* be a *perfect* elimination ordering is when `G` *is* chordal! So returning a dict of 0's really makes no sense (hence the original issue I suspect).

IMO the thing that's needed most here is clarity on the definitions and semantics. From a blank slate (i.e. ignoring the current behavior), my expectation would be:

- `complete_to_chordal_graph` returns the graph `H` which was made chordal, and/or the set of edges that were added to the input graph `G` to make `H` chordal.
- `perfect_elimination_ordering` returns (or yields) a sequence of nodes that, if they were to be removed from `G` in that order, would result in a chordal graph at each removal.

IMO it's important to decide what the correct path forward is rather than using the existing behavior as constraints for new functionality.

**Comment by Luis-Varona** [comment]
@rossbar, as per the added documentation in PR #8024, a *perfect elimination ordering* (or PEO) is an ordering of the vertices of a chordal graph *G* such that for every vertex *v* in *V(G)*, the set of neighbors of *v* that occur later in the PEO form a clique. On the other hand, a *minimal elimination ordering* (or MEO) is an ordering of the vertices of a graph *G* such that after *G* has been *minimally triangulated* to a chordal graph *H*, the MEO is a PEO for *H*. (When *G* is already chordal, then we simply have *H = G*.)

`complete_to_chordal_graph(G)` currently returns a minimal triangulation *H* of *G* (that is, a chordal graph created by adding the minimum number of vertices to *G*) as well as a PEO for this new graph *H* (in other words, an MEO for *G*). When *G* is already chordal, however, `complete_to_chordal_graph(G)` skips the computation of the MEO, simply returning all zeros.

Hence, the approach I've taken in PR #8024 (just pending review from @dschult) is to add the functions `minimal_elimination_order` and `perfect_elimination_order`. `minimal_elimination_order(G)` does the same thing as `complete_to_chordal_graph(G)`, except even when *G* is already chordal, it calls the MCS-M triangulation algorithm to compute an MEO for *G* anyway. `perfect_elimination_order(G)` assumes that the input *G* is already chordal, hence raising an error if it is not.

This is because a graph is chordal if and only if it has a PEO, so if our input graph is not chordal, calling `perfect_elimination_order` makes no sense. On the other hand, every graph has an MEO (in the case of chordal graphs, an MEO is also a PEO), so calling `minimal_elimination_order` always makes sense.

I do see merit in the point that naming conventions could be improved to indicate that the two new functions not only return the MEO/PEO but also the triangulated (chordal) graph, but I hope this clears things up somewhat. Your thoughts, @dschult?

**Comment by Luis-Varona** [comment]
@dschult @rossbar - what should we do next? (Just revisiting this issue after quite a while.)

**Comment by dschult** [comment]
I think the discussion in #8024 seems to converge to an interface with 3 public functions:
- `is_chordal` which I think already exists
- `chordal_triangulation_edges` which returns a list of edges needed to make G chordal
- `perfect_elimination_order` (or maybe `perfect_elimination_ordering`) which returns a list of nodes.

I think the next steps are to merge the PR with main and resolve any conflicts, make the code changes reflect the API we converged to, adjust tests and docs.  Does this make sense? Are there any remaining big picture questions that you see?
